mipengine/controller/node_landscape_aggregator.py

```diff
@@ -0,0 +1,287 @@
+            data_model_registry.set_datasets_location(datasets_locations)
+            logger.debug(f"Nodes:{[node.id for node in node_registry.nodes]}")
+            # ..to print full nodes info
+            # from devtools import debug
+            # debug(self.nodes)
+            # DEBUG end
+
+            sys.stdout.flush()
+            await asyncio.sleep(NODE_LANDSCAPE_AGGREGATOR_UPDATE_INTERVAL)
+
+
+async def _get_datasets_locations(nodes):
+    datasets_locations = {}
+    for node_info in nodes:
+        node_socket_addr = _get_node_socket_addr(node_info)
+        datasets_per_data_model = await _get_node_datasets_per_data_model(
+            node_socket_addr
+        )
+        for data_model, datasets in datasets_per_data_model.items():
+            if data_model not in datasets_locations:
+                datasets_locations[data_model] = {dataset: [] for dataset in datasets}
+            for dataset in datasets:
+                datasets_locations[data_model][dataset].append(node_info.id)
+    return datasets_locations
+
+
+async def _get_datasets_labels(nodes):
+    datasets_labels = {}
+    for node_info in nodes:
+        node_socket_addr = _get_node_socket_addr(node_info)
+        datasets_per_data_model = await _get_node_datasets_per_data_model(
+            node_socket_addr
+        )
+        for data_model, datasets in datasets_per_data_model.items():
+            datasets_labels[data_model] = {}
+            for dataset in datasets:
+                datasets_labels[data_model][dataset] = datasets[dataset]
+    return datasets_labels
+
+
+async def _get_cdes_across_nodes(nodes):
+    nodes_cdes = {}
+    for node_info in nodes:
+        node_socket_addr = _get_node_socket_addr(node_info)
+        datasets_per_data_model = await _get_node_datasets_per_data_model(
+            node_socket_addr
+        )
+        for data_model in datasets_per_data_model:
+            cdes = await _get_node_cdes(node_socket_addr, data_model)
+            if data_model not in nodes_cdes:
+                nodes_cdes[data_model] = []
+            nodes_cdes[data_model].append((node_info.id, cdes))
+    return nodes_cdes
+
+
+def _get_common_data_models(
+    nodes_cdes: Dict[str, List[Tuple[str, CommonDataElements]]]
+) -> Dict[str, CommonDataElements]:
+    common_data_models = {}
+    for data_model, cdes_from_all_nodes in nodes_cdes.items():
+        first_node, first_cdes = cdes_from_all_nodes[0]
+        for node, cdes in cdes_from_all_nodes[1:]:
+            try:
+                first_cdes == cdes
+            except IncompatibleCDEs as exc:
+                logger.info(
+                    f"Node '{first_node}' and node '{node}' on data model '{data_model}' threw: {exc.message}"
+                )
+                break
+        else:
+            common_data_models[data_model] = first_cdes
+
+    return common_data_models
+
+
+def _get_common_data_models_with_valid_datasets(
```

Comment:
There is a lot of logic in this aggregator that I understand, but we won't remember in a year.

I think some documentation is needed in the form of method description at least. Another way is to add some big module documentation at the top, like in the udfgenerator.